rust/routee-compass-powertrain/src/routee/prediction/interpolation/interpolation_speed_grade_model.rs

```diff
@@ -0,0 +1,160 @@
+use std::path::Path;
+
+use crate::routee::prediction::prediction_model::PredictionModel;
+
+use routee_compass_core::{
+    model::traversal::traversal_model_error::TraversalModelError,
+    model::unit::{as_f64::AsF64, EnergyRate, EnergyRateUnit, Grade, GradeUnit, Speed, SpeedUnit},
+};
+use smartcore::{
+    ensemble::random_forest_regressor::RandomForestRegressor, linalg::basic::matrix::DenseMatrix,
+};
+
+use super::utils::{linspace, BilinearInterp};
+
+pub struct InterpolationSpeedGradeModel {
+    interpolator: BilinearInterp,
+    speed_unit: SpeedUnit,
+    grade_unit: GradeUnit,
+    energy_rate_unit: EnergyRateUnit,
+}
+
+impl PredictionModel for InterpolationSpeedGradeModel {
+    fn predict(
+        &self,
+        speed: (Speed, SpeedUnit),
+        grade: (Grade, GradeUnit),
+    ) -> Result<(EnergyRate, EnergyRateUnit), TraversalModelError> {
+        let (speed, speed_unit) = speed;
+        let (grade, grade_unit) = grade;
+        let speed_value = speed_unit.convert(speed, self.speed_unit).as_f64();
+        let grade_value = grade_unit.convert(grade, self.grade_unit).as_f64();
+
+        // snap incoming speed and grade to the grid
+        let min_speed = self.interpolator.x[0].0;
+        let max_speed = self.interpolator.x[self.interpolator.x.len() - 1].0;
+        let min_grade = self.interpolator.y[0].0;
+        let max_grade = self.interpolator.y[self.interpolator.y.len() - 1].0;
+
+        let speed_value = speed_value.max(min_speed).min(max_speed);
+        let grade_value = grade_value.max(min_grade).min(max_grade);
+
+        let y = self
+            .interpolator
+            .interpolate(speed_value, grade_value)
+            .map_err(|e| {
+                TraversalModelError::PredictionModel(format!("Failed to interpolate: {}", e))
+            })?;
+
+        let energy_rate = EnergyRate::new(y);
+        Ok((energy_rate, self.energy_rate_unit))
+    }
+}
+
+impl InterpolationSpeedGradeModel {
+    #[allow(clippy::too_many_arguments)]
+    pub fn new<P: AsRef<Path>>(
+        routee_model_path: &P,
+        speed_unit: SpeedUnit,
+        speed_bounds: (Speed, Speed),
+        speed_bins: usize,
+        grade_unit: GradeUnit,
+        grade_bounds: (Grade, Grade),
+        grade_bins: usize,
+        energy_rate_unit: EnergyRateUnit,
+    ) -> Result<Self, TraversalModelError> {
+        // Load random forest binary file
+        let rf_binary = std::fs::read(routee_model_path).map_err(|e| {
+            TraversalModelError::FileReadError(
+                routee_model_path.as_ref().to_path_buf(),
+                e.to_string(),
+            )
+        })?;
+        // Load the random forest regressor so we can exercise it on a linear grid
+        let rf: RandomForestRegressor<f64, f64, DenseMatrix<f64>, Vec<f64>> =
```

Comment:
noticing that this is smartcore-only. is it possible to add an `underlying: ModelType` to `ModelType::Interpolate`:

```rust
pub enum ModelType {
    Smartcore,
    Onnx,
    Interpolate {
        underlying: ModelType,
        speed_lower_bound: Speed,
        speed_upper_bound: Speed,
        speed_bins: usize,
        grade_lower_bound: Grade,
        grade_upper_bound: Grade,
        grade_bins: usize,
    },
}
```

and then we can load + sample whatever model that is?

Reply:
yeah that's a great idea, I'll add that in